Articles/HowTos/CriticalBeginnerGitUsageTips.md

```diff
@@ -0,0 +1,99 @@
+But before pushing these commits to a remote shared branch, use [`git rebase -i @{u}`](https://www.atlassian.com/git/tutorials/rewriting-history#git-rebase-i) to clean up and reorganize the commits into good "logical" commits (see "SEPARATE CHANGES" above).
+
+**D) Always create local commits for your changes in local branches before you run any Git commands that might modify your local uncommitted changes.**
+The commands `git pull`, `git merge`, `git rebase`, or other Git operations can alter (or
+delete) your local uncommitted changes, either in the working directory or the staging
+area/index.
+So always create (sometimes temporary) commits for these before running any of these commands (unless you want to throw away uncommitted changes in your local repo, for example with `git checkout`, `git reset`, `git clean`, etc.).
+
+**E) Back up local branches every few hours of work by pushing them to some
+remote Git repo on a remote machine** in case your local machine or disk goes out.
+(This is why local Git branches are better than Git stashes because you can back them up to other repos in a version controlled way.
+You can’t do that with Git stashes.)
+
+**F) You can always recover an earlier state of any of your local branches** by
+looking at [`git reflog`](https://git-scm.com/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery) and then using a combination [`git checkout`](http://marklodato.github.io/visual-git-guide/index-en.html#checkout) and/or [`git reset –hard`](http://marklodato.github.io/visual-git-guide/index-en.html#reset), etc.
+(See [How to undo (almost) anything in Git](https://github.blog/2015-06-08-how-to-undo-almost-anything-with-git/#redo-after-undo-local).)
+
+**G) Never delete a local Git repo (unless you are completely done with it).**
+Your local Git repos have a [wealth of information that can’t be pushed to other Git repos](https://www.cs.cmu.edu/~davide/howto/git_lose.html) (e.g. `git rerere` info, `git reflog` info, etc.).
+Because of **'F'**, you should never have to delete a local Git repo and re-clone to get out of some "bad" state.
+Only a corrupted disk that corrupts the local Git DB history (which is extremely rare) should cause you to have to re-clone the local Git repo.
+(If you feel like you need to delete your local Git repo to get out of some bad state of a non-corrupted repo, then you have not yet learned the basics of Git well enough so keep learning!)
+
+**H) Don’t commit large generated (binary) files in a Git repo** .
+These get stuck in the Git history **forever** and can only be removed by "[filtering](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)" the Git repo (which is a very disruptive process for everyone involved).
+
+**I) Never do `git push -f` to a remote branch shared with other people unless
```

Comment:
As an aside, it would be interesting to measure how many, in ECP for example, actually use shared branches for work (I mean apart from the branches all the development is getting poured into). I have seen this rarely used.

Reply:
I have been on projects where shared branches were pretty common (at least where more than 1 person pushed and pulled to the same shared branch).  The real power of git is in the usage of shared topic branches.